src/libs/logging/src/file_rotate.rs

```diff
@@ -0,0 +1,266 @@
+            if p.has_root() || !parent.as_os_str().is_empty() {
+                let md = parent.metadata()?;
+                if !md.is_dir() {
+                    return Err(io::Error::from(io::ErrorKind::InvalidInput));
+                }
+            }
+        }
+
+        Ok(FileRotator {
+            path: path.as_ref().to_path_buf(),
+            file: None,
+            truncate: false,
+            written_size: 0,
+            rotate_size: default_rotate_size(),
+            rotate_keep: default_rotate_keep(),
+            reopen_check_interval: Duration::from_millis(1000),
+            next_reopen_check: Instant::now(),
+        })
+    }
+
+    /// Enable/disable log file truncation.
+    pub fn truncate(&mut self, truncate: bool) -> &mut Self {
+        self.truncate = truncate;
+        self
+    }
+
+    /// Set the threshold size to rotate log files.
+    pub fn rotate_threshold(&mut self, size: u64) -> &mut Self {
+        self.rotate_size = size;
+        self
+    }
+
+    /// Set number of rotated log files to keep.
+    pub fn rotate_count(&mut self, count: usize) -> &mut Self {
+        self.rotate_keep = count;
+        self
+    }
+
+    /// Set the interval to check for rotation.
+    pub fn check_interval(&mut self, interval: Duration) -> &mut Self {
+        self.reopen_check_interval = interval;
+        self
+    }
+
+    fn reopen_if_needed(&mut self) -> Result<()> {
+        // Basically, path.exists() is VERY slow on windows, so we just
+        // can't check on every write. Limit checking to a predefined interval.
+        // This shouldn't create problems neither for users, nor for logrotate et al.,
+        // as explained in the issue.
+        let now = Instant::now();
+        let path_exists = if now >= self.next_reopen_check {
+            self.next_reopen_check = now + self.reopen_check_interval;
+            self.path.exists()
+        } else {
+            // Pretend the path exists without any actual checking.
+            true
+        };
+
+        if self.file.is_none() || !path_exists {
+            // If the old file was externally deleted and we attempt to open a new one before
+            // releasing the old handle, we get a Permission denied on Windows. Release the handle.
+            self.file = None;
+            let file = OpenOptions::new()
+                .create(true)
+                .write(true)
+                .truncate(self.truncate)
+                .append(!self.truncate)
+                .open(&self.path)?;
+            self.written_size = file.metadata()?.len();
+            self.file = Some(LineWriter::new(file));
+        }
+
+        Ok(())
+    }
+
+    fn rotate(&mut self) -> Result<()> {
+        let _ = self.file.take();
+
+        for i in (1..=self.rotate_keep).rev() {
+            let from = self.rotated_path(i)?;
+            let to = self.rotated_path(i + 1)?;
+            if from.exists() {
+                // Ignore the error
```

Comment:
Should we?

Reply:
The `fuse()` are used to chain slog drivers in `create_logger()`, which will panic when encounters errors. So error is ignored to avoid panicking the whole process when failed to rotate old log files.

Reply:
That's for the discussion, which has inspired me to rethink the error handling model. I will introduce two models for error handling:
1) best effort mode: ignore all errors happened during logging
2) guaranteed mode: return all errors from the writer